text/0000-fields-in-traits.md

```diff
@@ -0,0 +1,800 @@
+expression must be an expression of the form `self(.F)*` where `F`
+represents some field name (possibly a fully qualified one, see
+below). In other words, `self.a.b.c` would be legal, but
+`(*self.a.b).c` would not; nor would `self.a[b]`. Furthermore, the
+lvalue expression must not contain any implicit derefs.
+
+These rules serve to ensure the following properties:
+
+1. the lvalue is located within the `self` value with some fixed pointer offset;
+2. borrowck can easily evaluate whether two field mappings are disjoint.
+
+In the future, we might consider various extensions:
+
+- **The ability to index into fixed-length arrays with a constant
+  index.** However, it would be best to couple that with a general
+  overhaul of constant evaluation (and probably an extension of
+  borrowck to understand expressions of this form more broadly).b
+- **The ability to deref.**  This was excluded so as not
+  to complicate field access in trait objects, but it could be that
+  traits with lvalues that require passing through a reference are
+  simply not considered object safe.
+  
+#### Disjointness rules
+
+The field mappings within an impl must be disjoint from other field
+mappings within that same impl, or within super-trait impls. More
+specifically, if there is some impl containing a field mapping like so:
+
+```rust
+impl<...> Trait<U,V> for T {
+   ...
+   field: lvalue,
+   ...
+}
+```
+
+then `lvalue` must be *disjoint* (defined below) from the lvalues
+found in all other field mappings within that impl. Furthermore,
+`lvalue` must be disjoint from the lvalues mapped in supertrait impls.
+
+Note that if a trait has two unrelated supertraits, the fields of those
+supertraits do not have to be disjoint. This can be used to form arbitrary
+disjointness relationships. For example, imagine that we have some type `Foo`
+that implements the trait `C` shown below:
+
+```rust
+trait A {
+    f: u32,
+    g: u32, // must be disjoint from f
+}
+
+trait B {
+    h: u32, // can overlap with f and g
+}
+
+trait C: A + B {
+    i: u32, // must be disjoint from f, g, and h
+}
+```
+
+As the comments suggest, `Foo` could map `A::f` and `B::h` to the same
+lvalue if it wanted to, but it must map `A::f`, `A::g`, and `C::i` to
+mutually disjoint lvalues.
+
+#### Definition of "disjoint"
+
+Two lvalues are considered "disjoint" if the borrow checker would
+allow `&mut` borrows of them simultaneously. This is typically true
+because they are based in different fields of the same struct.
+
+### Fully qualified field access
+
+Generally speaking, the lookup for fields will follow the same rules
+as methods: if the type itself defines an "inherent" field, that field
+is used, but otherwise we search for in-scope traits that implement
+the trait and define a field with that name. Sometimes it may be
+necessary or desirable to specify the trait explicitly. For those
+cases, we introduce a fully qualified field notation which looks like
+`x.<Trait<U,V>::f>`.
```

Comment:
`x` is not supposed to be a type, but an expression, so that form wouldn't necessarily be unambiguous.